File: packages/native/src/host/n8/RisaSyncTranslator.cpp
#include "RisaSyncTranslator.hpp"

namespace retroplug {

std::array<std::uint8_t, 5> RisaSyncTranslator::armPacket(std::int64_t absoluteClock) {
    if (absoluteClock < 0) absoluteClock = 0;
    const std::int64_t phrase   = absoluteClock / RISA_CLOCKS_PER_PHRASE;
    const std::uint8_t songRow  = static_cast<std::uint8_t>((phrase >> 4) & 0x7f);
    const std::uint8_t chainRow = static_cast<std::uint8_t>(phrase & 0x0f);
    const std::uint8_t tick     = static_cast<std::uint8_t>((absoluteClock % RISA_CLOCKS_PER_PHRASE) & 0x7f);
    return { RISA_LOCATE_STATUS, RISA_LOCATE_SUB, songRow, chainRow, tick };
}

void RisaSyncTranslator::arm(std::vector<std::uint8_t>& out) {
    // F9 52 ss cc tt (the arm is a barrier: risa discards the old position's queued clocks), then FA start.
    const std::array<std::uint8_t, 5> pkt = armPacket(absoluteClock_);
    out.insert(out.end(), pkt.begin(), pkt.end());
    out.push_back(RISA_START);
    playing_ = true;
    // risa performs one priming sequencer tick itself when it applies the locate, so the armed clock must
    // NOT also get an F8 (it would double-advance the row). Suppress exactly the next incoming clock.
    suppressNextClock_ = true;
}

void RisaSyncTranslator::onMessage(const std::uint8_t* bytes, std::size_t n, std::vector<std::uint8_t>& out) {
    if (n == 0) return;
    const std::uint8_t status = bytes[0];

    switch (status) {
        case MIDI_SPP:
            // Song Position Pointer: F2 lsb msb, a 14-bit position in MIDI beats (sixteenth notes). Each
            // sixteenth is six 24-PPQN clocks. Sets the position a subsequent Continue arms from.
            if (n >= 3) {
                const std::int64_t pos16th = (static_cast<std::int64_t>(bytes[2] & 0x7f) << 7) |
                                             static_cast<std::int64_t>(bytes[1] & 0x7f);
                absoluteClock_ = pos16th * 6;
            }
            return;

        case MIDI_START:
            // MIDI Start is defined as "from the top" - arm at clock 0 regardless of any prior position.
            absoluteClock_ = 0;
            arm(out);
            return;

        case MIDI_CONTINUE:
            // Continue resumes from the current position (last SPP, or where Stop left the playhead).
            arm(out);
            return;

        case MIDI_STOP:
            // Gate + stop. Keep absoluteClock_ so a following Continue resumes in place.
            out.push_back(RISA_STOP);
            playing_ = false;
            suppressNextClock_ = false;
            return;

        case MIDI_CLOCK:
            if (!playing_) return;  // clocks outside transport are ignored (matches the role's transport gate)
            // Advance the position on every received tick so a later re-arm's locate stays aligned, but
            // suppress emitting the one armed clock that risa primed itself.
            ++absoluteClock_;
            if (suppressNextClock_) { suppressNextClock_ = false; return; }
            out.push_back(RISA_CLOCK);
            return;

        default:
            // Notes / CC / sysex / other real-time (active-sensing, reset): risa is transport-driven, ignore.
            return;
    }
}

}  // namespace retroplug

File: packages/retroplug/src/n8/risaSyncTranslator.ts
// MIDI clock/transport -> risa host-sync bytes. A pure, I/O-free translator that turns an incoming MIDI
// transport stream (SPP / Start / Continue / Stop / Clock) into risa's arm/clock/stop protocol. It reuses
// the risaLocate/risaArmPacket primitives from ../risaSync (the single source of truth) - this replaces the
// C++ RisaSyncTranslator, which was a hand-port of the same logic. Feed one MIDI message at a time via
// onMessage(); it appends any risa output bytes. Cross-checked against the golden role test
// (test/dsp/risa-sync.test.ts) and the former native gtest.

import { RISA_START, RISA_CLOCK, RISA_STOP, risaLocate, risaArmPacket } from "../risaSync";

const MIDI_SPP = 0xf2;
const MIDI_CLOCK = 0xf8;
const MIDI_START = 0xfa;
const MIDI_CONTINUE = 0xfb;
const MIDI_STOP = 0xfc;

export class RisaSyncTranslator {
  private absoluteClock = 0; // 24-PPQN clock since song start; set by SPP, advanced per F8
  private playingFlag = false;
  private suppressNextClock = false; // skip one F8 after a start (risa primes the armed clock itself)

  /** The 5-byte arm/locate packet for a 24-PPQN clock: F9 52 songRow chainRow tick. */
  static armPacket(absoluteClock: number): number[] {
    return risaArmPacket(risaLocate(absoluteClock / 24)); // risaLocate re-multiplies by 24 -> the same clock
  }

  playing(): boolean {
    return this.playingFlag;
  }
  getAbsoluteClock(): number {
    return this.absoluteClock;
  }

  /** Consume one MIDI message, appending any risa bytes to `out` (does NOT clear it). */
  onMessage(bytes: ArrayLike<number>, out: number[]): void {
    if (bytes.length === 0) return;
    switch (bytes[0]) {
      case MIDI_SPP: {
        if (bytes.length < 3) return;
        const pos16th = ((bytes[2] & 0x7f) << 7) | (bytes[1] & 0x7f);
        this.absoluteClock = pos16th * 6; // six 24-PPQN clocks per sixteenth; emits nothing
        return;
      }
      case MIDI_START:
        this.absoluteClock = 0; // always from the top
        this.arm(out);
        return;
      case MIDI_CONTINUE:
        this.arm(out); // arm at the current position (last SPP, or where Stop left it)
        return;
      case MIDI_STOP:
        out.push(RISA_STOP);
        this.playingFlag = false;
        this.suppressNextClock = false; // keep absoluteClock so a later Continue resumes in place
        return;
      case MIDI_CLOCK:
        if (!this.playingFlag) return; // transport-gated
        this.absoluteClock++; // advance FIRST so a later re-arm stays aligned
        if (this.suppressNextClock) {
          this.suppressNextClock = false;
          return; // the armed clock is primed by risa - drop this one
        }
        out.push(RISA_CLOCK);
        return;
      default:
        return; // notes / CC / sysex / other real-time ignored
    }
  }

  private arm(out: number[]): void {
    for (const b of RisaSyncTranslator.armPacket(this.absoluteClock)) out.push(b);
    out.push(RISA_START);
    this.playingFlag = true;
    this.suppressNextClock = true;
  }
}
